### PR TITLE
[REVIEW] Adding support for  `cudf.testing` 

### LINKED ISSUE #5539: [FEA] I wish I could use cuDF to compare DataFrames as in Pandas
**Is your feature request related to a problem? Please describe.**

I wish I could use cuDF to compare DataFrames as in pandas.

**Describe the solution you'd like**

An implementation of `assert_frame_equal()` method

**Describe alternatives you've considered**

Right now we compare as `all(dataFrame1 == dataFrame2)`

**Additional context**

Pandas example:
```python
import pandas as pd
data = pd.DataFrame(
data=([[1, 10, 100, 1000, 10000], [2, 20, 200, 2000, 20000]]), columns=['a', 'b', 'c', 'd', 'e']
      )
pd.testing.assert_frame_equal(data, data)

```

cuDF example:
```python
import cudf as pd

data = pd.DataFrame(
data=([[1, 10, 100, 1000, 10000], [2, 20, 200, 2000, 20000]]), columns=['a', 'b', 'c', 'd', 'e']
      )
pd.testing.assert_frame_equal(data, data) # no method
```

Traceback:
```python
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-8-c0090b488ab1> in <module>
      5 data=([[1, 10, 100, 1000, 10000], [2, 20, 200, 2000, 20000]]), columns=['a', 'b', 'c', 'd', 'e']
      6       )
----> 7 pd.testing.assert_frame_equal(data, data) # no method

AttributeError: module 'cudf' has no attribute 'testing'
```

### LINKED ISSUE #5586: [FEA] cudf equivalent of pandas.testing.assert_frame_equal
**Is your feature request related to a problem? Please describe.**

I would like to write unit tests for cudf data frames. 

**Describe the solution you'd like**

It would be nice to have a cudf equivalent of `pandas.testing.assert_frame_equal` to unit testing purposes. 

If there's another standard procedure that I'm not aware of, please let me know. 

**Describe alternatives you've considered**

I'm able to sufficiently work around this using something like the below:

```
import cudf

sources1 =      [10, 11, 12]
destinations1 = [11, 12, 10]
weights1 =      [77, 88, 99]
df1 = cudf.DataFrame({'source': sources1, 'destination': destinations1, 'weight': weights1}).set_index('source')


sources2 =      [11, 12, 10]
destinations2 = [12, 10, 11]
weights2 =      [88, 99, 77]
df2 = cudf.DataFrame({'source': sources2, 'destination': destinations2, 'weight': weights2}).set_index('source')


print('\n\n', df1)
print('\n\n', df2)


assert len(df1.merge(df2, how="outer")) == len(df1)
assert len(df1.merge(df2, how="outer")) == len(df2)
```

**Additional context**

No additional context.

### LINKED ISSUE #5587: [FEA] Add support for `cudf.testing`
**Is your feature request related to a problem? Please describe.**
Add support for `cudf.testing` similar to `pandas.testing` and support following APIs

- [ ] cudf..testing.assert_frame_equal
- [ ] cudf.testing.assert_series_equal
- [ ] cudf..testing.assert_index_equal

### PR DESCRIPTION
**Pull request by rgsl888prabhu**
Added Doc and test cases to validate. Almost completed, sweeping through any corner cases.
closes #5587
closes #5539 
closes #5586

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: python/cudf/cudf/testing/testing.py [hunk 1]
```diff
@@ -0,0 +1,600 @@
...
+        exact=check_index_type,
+        check_names=check_names,
+        check_less_precise=check_less_precise,
+        check_exact=check_exact,
+        check_categorical=check_categorical,
+        obj=f"{obj}.columns",
+    )
+
+    if by_blocks:
+        raise NotImplementedError("by_blocks is not supported")
```
Comment (shwina) [inline python/cudf/cudf/testing/testing.py:589]: A `NotImplementedError` indicates that it may be implemented in the future, let's just ignore whatever value is provided.

**Comment by shwina** [comment]
For self-consistency, do you think we should now use `cudf.testing` in our own test utilities, i.e., https://github.com/rapidsai/cudf/blob/branch-0.15/python/cudf/cudf/tests/utils.py#L6?

**Comment by rgsl888prabhu** [comment]
> For self-consistency, do you think we should now use `cudf.testing` in our own test utilities, i.e., https://github.com/rapidsai/cudf/blob/branch-0.15/python/cudf/cudf/tests/utils.py#L6?

I like the idea, but we might have to convert pandas object to cudf in several cases while comparing and this might cause/hide issues, that;s only my concern.

Apart from that, this will resolve several issue such as handling null/NaN and others. But, any changes with pandas might hide which we would have normally found out with any test-case failures.

**Comment by shwina** [comment]
Makes sense!

File: python/cudf/cudf/testing/testing.py [hunk 2]
```diff
@@ -0,0 +1,581 @@
...
+    raise AssertionError(msg)
+
+
+def _check_types(
+    left, right, check_categorical=True, exact="equiv", obj="Index"
+):
+    if not exact or exact == "equiv":
+        if (
+            isinstance(left, cudf.RangeIndex)
+            and isinstance(right, cudf.Int64Index)
```
Comment (shwina) [inline python/cudf/cudf/testing/testing.py:42]: Unlike Pandas, we have `Int8Index`, `Int16Index`, etc., and we probably also want to include those here.

File: python/cudf/cudf/testing/testing.py [hunk 3]
```diff
@@ -0,0 +1,581 @@
...
+        )
+
+    if (
+        exact
+        and not isinstance(left, cudf.MultiIndex)
+        and is_categorical_dtype(left)
+    ):
+        if left.dtype != right.dtype:
+            raise_assert_detail(
+                obj, "Catgorical difference", f"{left}", f"{right}"
```
Comment (shwina) [inline python/cudf/cudf/testing/testing.py:61]:
```suggestion
                obj, "Categorical difference", f"{left}", f"{right}"
```

File: python/cudf/cudf/testing/testing.py [hunk 4]
```diff
@@ -0,0 +1,581 @@
...
+        and is_categorical_dtype(left)
+        and is_categorical_dtype(right)
+    ):
+        left = left.astype(left.categories.dtype)
+        right = right.astype(right.categories.dtype)
+
+    columns_equal = False
+    try:
+        columns_equal = left.equals(right)
+    except TypeError:
```
Comment (shwina) [inline python/cudf/cudf/testing/testing.py:176]: Can we document what we're expecting to raise this `TypeError` as it's not clear

Comment (rgsl888prabhu) [reply python/cudf/cudf/testing/testing.py:176]: When we compare to two categorical columns, and if the categories don't match it will throw TypeError, I will also update it with a comment.

Comment (shwina) [reply python/cudf/cudf/testing/testing.py:176]: OK. The logic is a bit hard to follow here.

```
    try:
        columns_equal = left.equals(right)
    except TypeError:
        if is_categorical_dtype(left) and is_categorical_dtype(right):
            left = left.astype(left.categories.dtype)
            right = right.astype(right.categories.dtype)
```

What if a `TypeError` is raised and `left` and `right` are *not* categoricals? Then we're effectively ignoring that `TypeError`. Is that the desired behaviour? Or should we re-raise?

Comment (rgsl888prabhu) [reply python/cudf/cudf/testing/testing.py:176]: We should re-raise, I will make sure this is taken care only in the case I have described.
